fix misuse of LLVM API found through inkwell tests

nlewycky opened the pull request:
Running inkwell's tests with against an asserts-enabled expensive-checks-enabled, address-sanitizer undefined-behaviour-sanitizer LLVM found some problems. The ones I have fixes for are collected in this branch, the more complicated ones I filed (or will file) bugs for.

This is not yet enough to make inkwell tests run cleanly on my test system, this contains only the parts I think are clean enough to be committed. Most of the changes are to the tests rather than deeper changes that make inkwell safe. I want to work on that too, but I'm going to start with just getting to green (tests passing).

## Description

I've marked the changes where I only fixed the test but where inkwell still allows a consumer to trigger the problem.

 * const_null() is removed. See \<Breaking Changes\>.
 * the element type of a vector may only be an integer, floating point, or pointer.  [test only]
 * void type can not be used as an aggregate in a pointer, structure, array, etc.  [test only]
 * LLVMConstIntGet[SZ]ExtValue are convenience functions to access the value of your constant integer when it is 64 bits or less. Return None for larger integers.
 * test_enum_attributes() is removed, it was testing creation and use of custom kinds of enum attributes. As far as I can tell, LLVM does not intend to allow library users to create their own kinds of attributes, and it sometimes asserts on it.  [test only]
 * don't build a bitcast from f32 to i64. LLVM will detect that this bitcast can't be valid and assert. (In general, LLVM tries hard to assert as early as possible with the intention that this will help developers find bugs.)   [test only]
 * 12 is not a valid alignment as it is not a power of 2. Replaced with 16.  [test only]
 * fix 8-bit literals in right shift example in doc string vs. rust nightly.

## Related Issue

Related to issues #73 and #80.

This is part of what I mentioned was coming in PR #92.

## How This Has Been Tested

`cargo +nightly test --features=llvm7-0` but also I did a build of LLVM 7.0.1 with ASan and UBSan and added RUSTFLAGS="-Z sanitizer=address" to the test run.

## Option\<Breaking Changes\>

I deleted the const_null() function. It does exactly what const_zero() does except when it exhibits undefined behaviour.

## Checklist

- [x] I have read the [Contributing Guide](https://github.com/TheDan64/inkwell/blob/master/.github/CONTRIBUTING.md)

TheDan64 commented:
Could you explain this some more:
> I deleted the const_null() function. It does exactly what const_zero() does except when it exhibits undefined behavior.

Why/when is it UB? Why isn't `const_zero`?

TheDan64 commented:
> don't build a bitcast from f32 to i64. LLVM will detect that this bitcast can't be valid and assert. (In general, LLVM tries hard to assert as early as possible with the intention that this will help developers find bugs.) [test only]

I wonder if the C api has a way for us to compare sizes ourselves...

> 12 is not a valid alignment as it is not a power of 2. Replaced with 16. [test only]

We should probably assert that `set_param_alignment_attribute`'s arg is a power of two (ie `num::count_ones() == 1`?)

TheDan64 commented on src/values/int_value.rs:
```diff
@@ -444,6 +447,9 @@ impl IntValue {
         if !self.is_const() {
             return None;
         }
+        if self.get_type().get_bit_width() > 64 {
```
> LLVMConstIntGet[SZ]ExtValue are convenience functions to access the value of your constant integer when it is 64 bits or less. Return None for larger integers.

LLVM supports 128bit ints as well as custom bit width ints. Why wouldn't this work for > 64?

nlewycky replied:
> > LLVMConstIntGet[SZ]ExtValue are convenience functions to access the value of your constant integer when it is 64 bits or less. Return None for larger integers.
> 
> LLVM supports 128bit ints as well as custom bit width ints. Why wouldn't this work for > 64?

On common platforms the largest type available in C and C++ is a 64-bit type, so LLVM provided a convenience function that returns a `uin64_t`. (This is akin to the fact that llvm supports 256-bit integers but rust only has a `u128`.) If the value is larger, LLVMConstIntGet[SZ]ExtValue will saturate. In C++ you can access the larger value with ConstantInt::getValue(), but there doesn't appear to be any way at all to get it through the C API.

TheDan64 commented on src/builder.rs:
```diff
@@ -834,14 +834,14 @@ impl Builder {
-    /// assert_eq!(0b1110_0000i8 >> 1, 0b1111_0000);
-    /// assert_eq!(0b1100_0000i8 >> 2, 0b1111_0000);
+    /// assert_eq!(0b1110_0000u8 as i8 >> 1, 0b1111_0000u8 as i8);
+    /// assert_eq!(0b1100_0000u8 as i8 >> 2, 0b1111_0000u8 as i8);
```
Hmm, how come the cast is necessary?

nlewycky replied:
Looks like a change in rust nightly?

TheDan64 commented on tests/all/test_builder.rs:
```diff
@@ -695,7 +698,7 @@ fn test_bitcast() {
+    builder.build_bitcast(f64_arg, i64_type, "f64toi64");
 
-    assert!(module.verify().is_err());
+    assert!(module.verify().is_ok());
```
I think the point of this was to test that LLVM will infact declare the module invalid if you try to bitcast `f32` to `i64`. So long as it isn't UB, what's the issue with this?

nlewycky replied:
An asserts build of LLVM will assert and terminate the program.

It's true that creating the bitcast instruction won't trigger UB internally to LLVM, but there is no way to ever convert such a bitcast into code that in valid. The type of a value is immutable assigned only upon creation, so neither the target type of the bitcast nor the type of its operand can change. LLVM has a similar assertion for RAUW between different types too.

nlewycky commented:
> Could you explain this some more:
> 
> > I deleted the const_null() function. It does exactly what const_zero() does except when it exhibits undefined behavior.
> 
> Why/when is it UB? Why isn't `const_zero`?

It casts the Type* it's given to a PointerType*. If the target memory is not a PointerType, that's UB on the spot. In an asserts build of llvm, it's a checked cast with an assertion failure. In non-asserts build of llvm, it's just an unchecked cast to an object of the wrong type, then it starts dereferencing members that don't exist on PointerType.

> > don't build a bitcast from f32 to i64. LLVM will detect that this bitcast can't be valid and assert. (In general, LLVM tries hard to assert as early as possible with the intention that this will help developers find bugs.) [test only]
> 
> I wonder if the C api has a way for us to compare sizes ourselves...
> 
> > 12 is not a valid alignment as it is not a power of 2. Replaced with 16. [test only]
> 
> We should probably assert that `set_param_alignment_attribute`'s arg is a power of two (ie `num::count_ones() == 1`?) We should also mention this in the method's docs.

There's LLVMGetIntTypeWidth but only if the type is an IntegerType, LLVMGetVectorSize for VectorType, there's LLVMSizeOf which produces a Value that can be used at runtime (but we could check if it's a constant int and pull its value?). Ideally LLVM's CastInst::castIsValid would be exposed through the C API.
